### ISSUE TITLE
codegen / logic issue for "GIXSQLCursorDeclareParams"

### ISSUE THREAD

**Issue by GitMensch** [mridoni/gixsql#88]
A program using cursors get the following generated directly before the first section (I hope it isn't before / in `DECLARATIVES` if those exist);
```
*   ESQL STARTUP DECLARATIONS (START)
....
"GIXSQLStartSQL"
"GIXSQLSetSQLParams"   *> N times
"GIXSQLCursorDeclareParams"
"GIXSQLEndSQL"

... for each cursor

*   ESQL STARTUP DECLARATIONS (END) 
```

The issues here are:
* this happens for all cursors - while the program may not use any of those for nearly all calls
* this happens for _each_ invocation of the program and after the first iteration the cursors are setup leading the generated `_gixsqlCursorDeclare` always ending in `DBERR_CURSOR_EXISTS` and SQLCA being updated with error codes.

Suggestion to solve this:
* generated a variable with the cursor-name + "initialized" for each cursor in the program
* before each use of the cursor in the program: check if that variable is space - when yes set it to "X" and do the initializations _only for this cursor_

If this takes "too long for the next version" (I personally see it as big drawback in GixSQL and therefore suggest to fix that "soon") then generate a `MOVE SQLCA TO SQLCA-GIXSQL-BACKUP` directly after `ESQL STARTUP DECLARATIONS (START)` and a `MOVE SQLCA-GIXSQL-BACKUP TO SQLCA-GIXSQL-BACKUP` directly before `ESQL STARTUP DECLARATIONS (END)`.

**Comment by GitMensch** [comment]
re enhancement: as long as the SQLCA is not back-up'd and restored around this it is an actual error because the values "magically change" between invocations and the programs run as if the last statement was an error.

Readjusting the generation to do the preparation only once and only directly before the cursor is used the first time would definitely be both more work an enhancement improving performance and the ability to debug actual errors.

**Comment by mridoni** [comment]
I have implemented what follows, let me know if it might work. It is all activated with a new preprocessor flag (`--smart-cursor-init`).

- This only applies to cursors declared in `WORKING-STORAGE`. All the other cursors have no special treatment
- A series of flags is defined in `WORKING-STORAGE` to keep track of cursor initialization
- Cursor initialization statements are still output (almost) just after `PROCEDURE DIVISION`, but they are delimited by their own paragraphs
- A paragraph definition is output just after the last cursor initialization.
- A `GO TO` statement is output between `PROCEDURE DIVISION` and the first cursor initialization, that skips all the cursor initializations, going to the paragraph above
- Before each `OPEN`, `FETCH` or `CLOSE` SQL statement the initialization flag is checked. If it is not set, cursor initialization is performed with a `PERFORM THRU` to the appropriate paragraph, the flag is set and control resumes.

You can look at the attached (preprocessed) program, it is the same code as in #81, only preprocessed with the `--smart-cursor-init` flag, it compiles and runs successfully.

[TSQL027A.cbsql.txt](https://github.com/mridoni/gixsql/files/9070110/TSQL027A.cbsql.txt)

**Comment by GitMensch** [comment]
That seems very reasonable - I'd drop the `-EX` paragraphs and the `-ST` suffix, then just `PERFORM GIXSQL-CI-P-TSQL027A-VM3`.
And for some strange reason I'd place the `MOVE 'X' TO GIXSQL-CI-F-TSQL027A-VM3` into `GIXSQL-CI-P-TSQL027A-VM3` itself.

I further suggest to move the complete block to the end of the processed programm - this way you do not have that `GO TO` executed each time (visible by `-ftrace-all` and when using the debugger). This also simplifies the codegen as you do not have to look out for `DECLARATIVES` at all (at least not for gixsql) _and_ you'd comply with the COBOL standard rule "if any paragraph is in a section, every paragraph must" by  implicit inserting them into the last section if there is one.

**Comment by GitMensch** [comment]
You _may_ want to insert the call twice as follows:

```cobol
GIXSQL*    EXEC SQL AT :DBS
GIXSQL*        OPEN VM2 
GIXSQL*    END-EXEC.
GIXSQL     IF GIXSQL-CI-F-TSQL027A-VM2 = ' ' THEN
GIXSQL         PERFORM GIXSQL-CI-P-TSQL027A-VM2
GIXSQL         IF SQLCODE = 0
GIXSQL           MOVE 'X' TO GIXSQL-CI-F-TSQL027A-VM2
GIXSQL           CALL STATIC "GIXSQLCursorOpen" USING
GIXSQL               BY REFERENCE SQLCA
GIXSQL               BY REFERENCE "TSQL027A_VM2" & x"00"
GIXSQL           END-CALL
GIXSQL        END-IF
GIXSQL     ELSE
GIXSQL        CALL STATIC "GIXSQLCursorOpen" USING
GIXSQL            BY REFERENCE SQLCA
GIXSQL            BY REFERENCE "TSQL027A_VM2" & x"00"
GIXSQL        END-CALL
GIXSQL     END-IF
```
--> marker is only set if the initialization worked and the cursor access is only done if it did, too.

**Comment by mridoni** [comment]
> That seems very reasonable - I'd drop the `-EX` paragraphs and the `-ST` suffix, then just `PERFORM GIXSQL-CI-P-TSQL027A-VM3`.

I agree, it is simpler and more readable.

> And for some strange reason I'd place the `MOVE 'X' TO GIXSQL-CI-F-TSQL027A-VM3` into `GIXSQL-CI-P-TSQL027A-VM3` itself

In this case I wanted to avoid a jump, it might be more costly than checking the variable before (if needed) jumping, but I guess a lot depends on compiler optimizations. 

> I further suggest to move the complete block to the end of the processed programm

I am not sure about this : if there is no `STOP RUN`, `EXIT PROGRAM` or `GOBACK` before this initialization section, it will be executed, or am I missing something obvious?

**Comment by mridoni** [comment]
Done, this is the new preprocessed output.

[TSQL027A.cbsql.txt](https://github.com/mridoni/gixsql/files/9071290/TSQL027A.cbsql.txt)

**Comment by GitMensch** [comment]
Rechecked with 2 PCs - the new preprocessed output file is identical to the old one.

**Comment by mridoni** [comment]
No, "I" need a check, because I uploaded the wrong file, sorry :-)

[TSQL027A-2.cbsql.txt](https://github.com/mridoni/gixsql/files/9071332/TSQL027A-2.cbsql.txt)

**Comment by GitMensch** [comment]
`GIXSQL-CI-P-TSQL027A-VM2-ST` is generated before OPEN, FETCH and CLOSE - shouldn't it be just before the `OPEN` (not sure, so asking)?

You may want to indent the lines within `IF GIXSQL-CI-F-TSQL027A-VM2 = 'X' THEN` and may want to add at least one generated comment line before `GIXSQL*   ESQL CURSOR DECLARATIONS (START)` and `GIXSQL*   ESQL CURSOR INIT FLAGS (START)`, but other than that it looks good to me.

**Comment by mridoni** [comment]
> `GIXSQL-CI-P-TSQL027A-VM2-ST` is generated before OPEN, FETCH and CLOSE - shouldn't it be just before the `OPEN` (not sure, so asking)?

In theory, yes, but this would (probably) lead to a hard crash if the program flow is wrong and it - erroneously - performs a `FETCH` or a `CLOSE` before opening the cursor. This way, the cursor is always initialized, the `FETCH` fails, `SQLCODE` is set and hopefully intercepted, so the error in the program flow can be corrected.  But I have doubts about it too, since it could slow down `FETCH` operations too much (e.g. in a tight loop) and a badly coded program can be expected to crash anyway. Since this is the first implementation of this feature, for now I will comment the checks in `FETCH` and `CLOSE` generation, then will see if there is a need for them to keep things "cleaner".

> You may want to indent the lines within `IF GIXSQL-CI-F-TSQL027A-VM2 = 'X' THEN` and may want to add at least one generated comment line before `GIXSQL* ESQL CURSOR DECLARATIONS (START)` and `GIXSQL* ESQL CURSOR INIT FLAGS (START)`, but other than that it looks good to me.

Ok, done

**Comment by GitMensch** [comment]
> n theory, yes, but this would (probably) lead to a hard crash if the program flow is wrong and it - erroneously - performs a `FETCH` or a `CLOSE` before opening the cursor.

There should be a test case ensuring this does not happen, and from a quick glance I see no reason why it should:

https://github.com/mridoni/gixsql/blob/6b8ca51e912089acdf84a54c5d60301faf3f7f8f/runtime/libgixsql/gixsql.cpp#L658-L663

It sets SQLCA as expected.

> since it could slow down `FETCH` operations too much (e.g. in a tight loop)

Yes, that's the point.

Looking forward to 1.0.17 with upcoming fixes and features .I guess that may land somewhere next week?

**Comment by mridoni** [comment]
> Looking forward to 1.0.17 with upcoming fixes and features .I guess that may land somewhere next week?

Yes, that's the release date I am more or less planning.

**Comment by mridoni** [comment]
> Looks like the very good changes you already had for this did not get into 1.0.17?

Have you run gixpp woth `-L/smart-cursor-init`? This should be present and working.

**Comment by GitMensch** [comment]
No, of course not, as it isn't mentioned in https://github.com/mridoni/gixsql/releases/tag/v1.0.17 - is there a reason for making this optional (even more: not default)?

**Comment by GitMensch** [comment]
Note: the flag changes the codegen, but does not work, the compiler complains about a bunch of not existing variables and sections:

> TEST.cob:647: error: 'GIXSQL-CI-F-TEST-CSKEY01N' is not defined
TEST.cob:752: error: 'GIXSQL-CI-F-TEST-CSKEY01NG' is not defined
TEST.cob:912: error: 'GIXSQL-CI-F-TEST-CSKEY01B' is not defined
TEST.cob:1017: error: 'GIXSQL-CI-F-TEST-CSKEY01BG' is not defined
TEST.cob:648: error: 'GIXSQL-CI-P-TEST-CSKEY01N' is not defined
TEST.cob:753: error: 'GIXSQL-CI-P-TEST-CSKEY01NG' is not defined
TEST.cob:913: error: 'GIXSQL-CI-P-TEST-CSKEY01B' is not defined
TEST.cob:1018: error: 'GIXSQL-CI-P-TEST-CSKEY01BG' is not defined

**Comment by mridoni** [comment]
Ok, I will look into it again (and set "smart cursor init" as default). As you know my test passes, so there is probably something in the code you are testing with that exposes a bug not present in my own test code. I'll try to dig deeper. 

Thanks

**Comment by GitMensch** [comment]
> Ok, I will look into it again (and set "smart cursor init" as default)

That has exposed bugs for the VARLENGTH32 issue, too, so maybe you get a reproducer "for free".

**Comment by GitMensch** [comment]
I couldn't reproduce a broken `-L` on the first try (will try more), but found #94 during debugging (where I've found `-L` to generate reasonable working code).

**Comment by GitMensch** [comment]
Think I've found it: it happens if you have the cursor declarations within the program, not within `SQL EXEC DECLARE` (before the `PROCEDURE DIVISION`). Shouldn't your tests catch this as soon you change the default to be `-L`?

**Comment by mridoni** [comment]
Yes, they should, but they didn't... 😞 I will look into it, thanks for your investigation.

**Comment by GitMensch** [comment]
Another issue seen with cursor declarations in `WORKING-STORAGE`: 

The variable `GIXSQL-CI-F-PROG-CSKEY01NG` was added directly before `PROCEDURE DIVISION` - but it must be placed at the end of `WORKING-STORAGE` (not in `LINKAGE` as it is currently...).

**Comment by mridoni** [comment]
Found it: while the "init flags" were correctly emitted only for cursors declared in `WORKING-STORAGE`, the "init check" was emitted for all cursors, while it shouldn't have been. Obviously the corresponding init flag had not been generated and the preprocessed code could not be compiled.

This was compounded by the problem with the init flags being emitted in the wrong place when a `LINKAGE SECTION` is present. This has also been fixed.

I am re-running the regression tests (test case TSQL027 has been modified to account for these two cases), then I will probably issue a pre-release ("autoconf tarball" only) sometime in the afternoon.

**Comment by mridoni** [comment]
Re-checking

**Comment by GitMensch** [comment]
> Another issue seen with cursor declarations in `WORKING-STORAGE`:
> 
> The variable `GIXSQL-CI-F-PROG-CSKEY01NG` was added directly before `PROCEDURE DIVISION` - but it must be placed at the end of `WORKING-STORAGE` (not in `LINKAGE` as it is currently...).

This part is fixed with 1.0.18.dev1

**Comment by mridoni** [comment]
> Re-checking

Should be correct (with 1.0.18dev1):

- cursors declared in `WORKING-STORAGE` are initialized in a block emitted t the end of the code, initialization flags are emitted and checked before `OPEN`
- cursors declared in `PROCEDURE DIVISION` are initialized "inline" when an `EXEC SQL DECLARE...CURSOR` is encountered in the source, no special management is needed (was incorrectly provided in v1.0.17).

I am attaching for reference my (revised since 1.0.18dev1) test case, original and preprocessed
- [TSQL027A.cbl.txt](https://github.com/mridoni/gixsql/files/9211996/TSQL027A.cbl.txt)
- [TSQL027A.cbsql.txt](https://github.com/mridoni/gixsql/files/9211995/TSQL027A.cbsql.txt)

Thanks

**Comment by GitMensch** [comment]
I see clear now! Wouldn't it make sense to do the smart-init also for entries in `PROCEDURE DIVISION` (not this bug, which may be closed now, but a feature-request).

**Comment by mridoni** [comment]
> Just mentioning - I don't know if it is possible to use variables defined in `LINKAGE` (or declared as `BASED`) with gixsql (or any esql precompiler) for cursor access.

The ProCOB documentation doesn't mention anything about it, IBM says [here](https://www.ibm.com/docs/en/cobol-zos/6.1?topic=environment-coding-sql-statements):

 >Define all host variables that you use in SQL statements in the WORKING-STORAGE SECTION, LOCAL-STORAGE SECTION, or LINKAGE SECTION. However, you do not need to identify them with EXEC SQL BEGIN DECLARE SECTION and EXEC SQL END DECLARE SECTION.

**Comment by mridoni** [comment]
> I see clear now! Wouldn't it make sense to do the smart-init also for entries in `PROCEDURE DIVISION` (not this bug, which may be closed now, but a feature-request).

I don't know if it might help with called programs, but this might be added anyway, since there is actually no harm in it.

**Comment by GitMensch** [comment]
> > I see clear now! Wouldn't it make sense to do the smart-init also for entries in `PROCEDURE DIVISION` (not this bug, which may be closed now, but a feature-request).
> 
> I don't know if it might help with called programs, but this might be added anyway, since there is actually no harm in it.

No harm, and not helping with called programs - only performance benefit.

For programs that have their parameters in `LINKAGE` (or as `BASED`) the parameter binding needs to be done _always_ as those may change at any time in the program. It _may_ be reasonable to have something like an internal check "is this still the same parameter address" and only go down to the driver if it is (for those another version of `GIXSQLSetSQLParams` may or may not be useful, something like `GIXSQLSetSQLParamsAgain`).

As the smart-init is new (solving both some performance-issues and an "older" bug of trashing SQLCA on program re-enter) [and the default] the "GIXSQLSetSQLParamsAgain" issue (however handled) should be done before the 1.0.18 release (I personally don't need it, but the release would be broken for `LINKAGE` and `BASED` fields otherwise).

**Comment by GitMensch** [comment]
Can't be closed in any case - because currently `GIXSQLCursorClose` calls are not changing the variable back :-/
The easiest option would likely to pass the `GIXSQL-CI-F` as additional parameter and if that call was successful, then set it to space. The other option would be to check the sqlcode in the program afterwards and set it there.

Because this isn't done the programs get a sqlstatus of -111 `DBERR_NO_SUCH_CURSOR` on its first cursor use after close.

**Comment by mridoni** [comment]
This could be solved in a different way and it is probably caused by a different "logic" bug: currently when a cursor is closed, it is not just marked as such, but its declaration is removed. Since there is no `UNDECLARE CURSOR`, at the time it seemed the logical thing to do, but in all evidence it is not, since prevents one from re-opening the cursor after it has been closed.

The solution might be just marking it as closed and destroying only the resources (if any) allocated when it is was opened, so leaving it ready for being re-opened.

**Comment by GitMensch** [comment]
> The solution might be just marking it as closed and destroying only the resources (if any) allocated when it is was opened, so leaving it ready for being re-opened.

Rechecking IBM and Oracle docs lead me to the same conclusion.

While this "internal undeclare cursor on cursor close" is a separate (even older issue) it is related. Should we still create a separate issue for that to track it and reference it  in the release notes?

**Comment by mridoni** [comment]
> While this "internal undeclare cursor on cursor close" is a separate (even older issue) it is related. Should we still create a separate issue for that to track it and reference it in the release notes?

Opened #98

**Comment by GitMensch** [comment]
Just to link those: #101 is related to this (nearly the same, but this one was about declarations in `DATA DIVISION`).

**Comment by mridoni** [comment]
Ok, I will look into it: it passed all the test cases, but it is really possible that this particular program flow eluded my tests.

Thanks

**Comment by GitMensch** [comment]
I'm still looking for the error at hand, the parameters seem correct as `GIXSQLCursorOpen` calls `DbInterfacePGSQL::cursor_open` which reads the parameters; that _would_ likely error though if the parameters were in `LINKAGE` because of the missing `GIXSQLSetSQLParams` (*only for parameters in `LINKAGE`!) the address is unchanged. I assume the other DB interfaces' cursor_open also read the parameters.

Minor thing I've seen during debugging: `cursor->setOpened()` should not be called from `GIXSQLCursorOpen` after it called `cursor_open` - as this already set the cursor state correctly.

**Comment by mridoni** [comment]
This could be solved with a bit of refactoring anyway, if necessary: for each cursor there are two paragraphs emitted in the initialization code at the end of the module:

- "paragraph 1" is generated,  named (as it is at the moment) `GIXSQL-CI-P-MYCRSR`. This would only contain the actual "declare" code 
- "paragraph 2" is generated, named - something like  `GIXSQL-CI-P-PAR-MYCRSR`. This would only contain the `StartSQL`/`SetSQLParams`/`EndSQL` calls
- when first opening a cursor, paragraph 2 is called, followed by paragraph 1
- during the subsequent "open" calls, the current check for paragraph 1 is performed, but paragraph 2 is *always* called

**Comment by GitMensch** [comment]
Sounds good. Note that if this is about the field information (type, length, address) only, then this is only necessary for items in `LINKAGE` (or defined with the `BASED` clause).

**Comment by mridoni** [comment]
> Sounds good. Note that if this is about the field information (type, length, address) only, then this is only necessary for items in `LINKAGE` (or defined with the `BASED` clause).

Yes, but at that point it would be more linear to behave like this in every case: the cost of doing it while opening a cursor, compared with the actual "open" on the DBMS side, is almost non-existent.

**Comment by GitMensch** [comment]
Retested: The codegen error with missing sections for cursor reads when the DECLARE is part of `PROCEDURE DIVISION` from 1.0.18,dev2 (or 1) is still there in 1.0.18 final. Can you reproduce this or do you need a reproducer (I could create one tomorrow)?

**Comment by mridoni** [comment]
Yes, a test case would be useful, since in all evidence the one I am using is not catching this.

Thanks

**Comment by GitMensch** [comment]
It always feels a bit strange if you take a big program and reduce it to the max for a minimal reproducer...

Here it is:

```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID.    TE609.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
      *-----------------------------------------------------------*
           EXEC SQL BEGIN DECLARE SECTION END-EXEC.
      *
       01 DBS                 PIC  X(010).
       01 TABID               PIC S9(018).
       01 MODTIME             PIC  X(026).

       77 KEY01-PART          PIC  9(008) VALUE 42.
      *
           EXEC SQL END DECLARE SECTION END-EXEC.
           EXEC SQL INCLUDE SQLCA END-EXEC.
      *-----------------------------------------------------------*
      *-----------------------------------------------------------*
       PROCEDURE DIVISION.
       TMAIN SECTION.
       P030-00.
      *
           EXEC SQL AT :DBS
              DECLARE CSKEY01N CURSOR FOR
                 SELECT TABID, MODTIME FROM TAB
                 WHERE
                    KEY01 >= (
                        LPAD(:KEY01-PART , 008, '0')
                             )
                 ORDER BY KEY01 ASC
           END-EXEC.
      *
           IF SQLCODE NOT = 0
              GO TO P031-99.
      *
           EXEC SQL
              OPEN CSKEY01N
           END-EXEC.
      *
           IF SQLCODE NOT = 0
              GO TO P031-99.
      *
       P031-20.
      *
           EXEC SQL
               FETCH CSKEY01N
               INTO
                 :TABID,
                 :MODTIME
           END-EXEC.

           EXEC SQL
              CLOSE CSKEY01N
            END-EXEC.
      *
       P031-99.
           GOBACK.
```

**Comment by mridoni** [comment]
Ok, thanks, I'll loot into it as soon as I can.

**Comment by mridoni** [comment]
Commenting out this line in the lexer makes it work, but even if it shouldn't be a problem, I will have (at least) to re-run all the tests for regressions.

https://github.com/mridoni/gixsql/blob/927cbcf2e169324b5d0f0ab9fd3ea9cc9cdaabb3/libgixpp/gix_esql_scanner.ll#L1296

**Comment by GitMensch** [comment]
`77` is definitely wrong here, as it is a plain variable with own storage.
`78` is a constant - so would be useful to be usable **as input** , but I don't know how  other preparsers handle this and if gixsql can handle "read only" already (actually should, because one can use "normal" literals already).
`88` items should not be used at all for `exec sql`, so dropping these seems reasonable (otherwise you could only error on use).
`66` items are "kind of redefines" with pre-fixed type (either the same, when there is only one entry, with `THROUGH` always as `PIC X` of appropriate size). I _guess_ that the parser _may_ not be able to handle those, in this case skipping them in the lexer seems reasonable (I don't think another sql preprocessor handles these).

**Comment by mridoni** [comment]
Ok, thanks. I will try removing `77` and `78` and leave the other two in then I will re-run all the tests.

**Comment by GitMensch** [comment]
> Ok, thanks. I will try removing `77` and `78` and leave the other two in then I will re-run all the tests.

... and best include a new test with both.

**Comment by GitMensch** [comment]
The separate issue is created, I suggest to let the scanner pass the level 78 if it is already in the parser and error out there, or just leave it as "skipped" (then the parser will error correctly with "unknown parameter"). It is fine to go with level 77 "as soon as possible", which seems (as no work other than the test case is missing) to be 1.0.18.a; if it is more work you may want to error in the parser for that, too.

**Comment by mridoni** [comment]
> It is fine to go with level 77 "as soon as possible", which seems (as no work other than the test case is missing) to be 1.0.18.a; if it is more work you may want to error in the parser for that, too.

The packages are being built, tomorrow I will run some more tests.

**Comment by GitMensch** [comment]
During testing with a patched 1.0.18 (including the possibly bad workaround for #119) I got "strange errors".
Debugging showed this may happen with code similar to the new testcase https://github.com/mridoni/gixsql/blob/5ed37df1f83f6fcbea04b66d1c1b6dbdd7e0cbce/gixsql-tests/data/TSQL017E.cbl#L24-L42

Currently the codegen is somewhat like

```cobol
GIXSQL*    EXEC SQL AT :DBS
GIXSQL*       DECLARE CRSR017D CURSOR FOR
GIXSQL*          SELECT TABID, MTIME FROM TAB
GIXSQL*          WHERE
GIXSQL*             KEY01 >= (
GIXSQL*                 MYFUNC(:KEY-ID-1 , :KEY-ID-2 , 099, '1')
GIXSQL*                      )
GIXSQL*          ORDER BY KEY01 ASC
GIXSQL*    END-EXEC.
      *
           IF SQLCODE NOT = 0
              GO TO P003-C.
      *
GIXSQL*    EXEC SQL
GIXSQL*       OPEN CRSR017D
GIXSQL*    END-EXEC.
      *
           IF SQLCODE NOT = 0
              GO TO P003-C.
```

It could be solved by adjusting the codegen for the "DECLARE in PROCEDURE DIVISION" case as follows (edit pending)

**Comment by mridoni** [comment]
Ok, I will look into it, but it might be a week or so until I find the time to work on it again, this is why I released v1.0.18a now. (I am also trying to release v1.0.18a for Gix-IDE, but today uploads on GitHub seem to be really really slow)

**Comment by GitMensch** [comment]
Sounds fair.

**Comment by GitMensch** [comment]
I'm aware that there's a lot ongoing from reading your dev-blog (mostly related to Gix Debugger and to packaging of Gix IDE)...
still I'd like to leave a ping asking about generation adjustments of SQLCODE.

Could you come up with a change for that (as it is still the reason the codebase in question - previously using ocesql - cannot be switched over to GixSQL) and release 1.0.18b, possibly with including #127?

**Comment by mridoni** [comment]
This patch modifies the code generation as indicated above. During the next few days (possibly the beginning of next week) I will apply the three patches (this one, the one for #119 and PR #127) and release v1.0.18b as a pre-release.

[issue88.patch.txt](https://github.com/mridoni/gixsql/files/9828211/issue88.patch.txt)
